### PR TITLE
[Feat] add tiny Autoencoder for (almost) instant decoding

### PR DESCRIPTION
**Pull request by sayakpaul**
Fixes https://github.com/huggingface/diffusers/issues/4233. 

## Considerations

I decided to NOT modify the `AutoencoderKL` class for this one because:

* The tiny Autoencoder is fundamentally trained differently. Refer to https://github.com/madebyollin/taesd for more details. 
* Very minimal resemblance to how we structure the blocks in `AutoencoderKL`. This tiny Autoencoder just has a bunch of conv + relu blocks with residual connections. 

### SD

```python
import torch
from diffusers import DiffusionPipeline, AutoencoderTiny

pipe = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-2-1-base", torch_dtype=torch.float16
)
pipe.vae = AutoencoderTiny.from_pretrained("sayakpaul/taesd-diffusers", torch_dtype=torch.float16)
pipe = pipe.to("cuda")

prompt = "slice of delicious New York-style berry cheesecake"
image = pipe(
    prompt, num_inference_steps=25, height=512, width=512, guidance_scale=3.0,
    generator=torch.manual_seed(0)
).images[0]
image
```

#### Comparison

| Tiny AE | Original AE |
|---|---|
|  |  |

### SDXL

```python
from diffusers import DiffusionPipeline, TinyAutoencoder
import torch

pipeline = DiffusionPipeline.from_pretrained(
    "stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float16
)
pipeline.vae = TinyAutoencoder.from_pretrained("sayakpaul/taesdxl-diffusers", torch_dtype=torch.float16)
pipeline = pipeline.to("cuda")

prompt = "slice of delicious New York-style berry cheesecake"
image = pipeline(
    prompt, num_inference_steps=25, guidance_scale=3.0,
    generator=torch.manual_seed(0)
).images[0]
image
```

| Tiny AE | Original AE |
|---|---|
|  |  |

## Todos

- [x] Add tests (will do once I get an initial review)
- [x] Add docs
- [ ] Transfer [sayakpaul/taesd-diffusers](https://huggingface.co/sayakpaul/taesd-diffusers) and [sayakpaul/taesdxl-diffusers](https://huggingface.co/sayakpaul/taesdxl-diffusers) to [madebyollin/taesd](https://huggingface.co/madebyollin/taesd) and [madebyollin/taesdxl](https://huggingface.co/madebyollin/taesdxl) respectively. 
- [ ] Can apply the changes (needed to accommodate for the tiny AE) in the affected pipelines such SD img-to-img, etc.

### CODE REVIEW

File: src/diffusers/models/tiny_autoencoder.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+        encoder_block_out_channels: Tuple[int] = (64, 64, 64, 64),
+        decoder_block_out_channels: Tuple[int] = (64, 64, 64, 64),
+        act_fn: str = "relu",
+        latent_channels: int = 4,
+        upsampling_scaling_factor: int = 2,
+        num_encoder_blocks: Tuple[int] = (1, 3, 3, 3),
+        num_decoder_blocks: Tuple[int] = (3, 3, 3, 1),
+        latent_magnitude: int = 3,
+        latent_shift: float = 0.5,
+        force_upcast: float = True,
```
[the remark this answers is not in the document]
Comment (madebyollin) [reply src/diffusers/models/tiny_autoencoder.py:198]: image

(Confirmed that TAESD activations should be in a wholesome `float16`-compatible range)

File: scripts/convert_tiny_autoencoder_to_diffusers.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+from diffusers import TinyAutoencoder
+
+
+"""
+Example - From the diffusers root directory:
+
+Download the weights:
+```sh
+$ wget -q https://huggingface.co/madebyollin/taesd/resolve/main/taesd_encoder.safetensors
+$ wget -q https://huggingface.co/madebyollin/taesd/resolve/main/taesd_decoder.safetensors
```
Comment (madebyollin) [inline scripts/convert_tiny_autoencoder_to_diffusers.py:20]: FYI, I will be AFK for a few days and won't be able to check `huggingface.co/madebyollin` PRs until end of the week

File: src/diffusers/image_processor.py [hunk 3]
```diff
@@ -110,11 +110,17 @@ def normalize(images):
...
-    def denormalize(images):
+    def denormalize(images, is_tiny_vae):
         """
         Denormalize an image array to [0,1].
+
+        Refer to https://github.com/madebyollin/taesd/issues/3#issuecomment-1657729279 to know why `is_tiny_vae`
+        exists.
         """
-        return (images / 2 + 0.5).clamp(0, 1)
+        if not is_tiny_vae:
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/image_processor.py:120]: We want to separate out the scaling logic from model, hence we have the `VaeImageProcessor`. So, that is why I think it appears to be more transparent way of doing it.

File: src/diffusers/models/tiny_autoencoder.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/models/tiny_autoencoder.py:198]: Thanks so much!

File: src/diffusers/models/tiny_autoencoder.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+        self.decoder = TinyDecoder(
+            in_channels=latent_channels,
+            out_channels=out_channels,
+            num_blocks=num_decoder_blocks,
+            block_out_channels=decoder_block_out_channels,
+            upsampling_scaling_factor=upsampling_scaling_factor,
+            act_fn=act_fn,
+        )
+
+        self.latent_magnitude = latent_magnitude
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/models/tiny_autoencoder.py:227]: I would be in favor of keeping it in case users want to experiment.

File: src/diffusers/models/tiny_autoencoder.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+    def unscale_latents(self, x):
+        """[0, 1] -> raw latents"""
+        return x.sub(self.latent_shift).mul(2 * self.latent_magnitude)
+
+    @apply_forward_hook
+    def encode(
+        self, x: torch.FloatTensor, return_dict: bool = True
+    ) -> Union[TinyAutoencoderOutput, Tuple[torch.FloatTensor]]:
+        output = self.encoder(x)
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/models/tiny_autoencoder.py:247]: We call the encode() method usually from training script where we ALWAYS ensure that the image pixel values are appropriately scaled :)

File: scripts/convert_tiny_autoencoder_to_diffusers.py (hunk 2 above)
Comment (sayakpaul) [reply scripts/convert_tiny_autoencoder_to_diffusers.py:20]: I don't think it will be blocker. 

Thanks for your awesome work, you rock!

File: src/diffusers/models/tiny_autoencoder.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
+# Copyright 2023 Ollin Boer Bohan and The HuggingFace Team. All rights reserved.
```
Comment (yiyixuxu) [inline src/diffusers/models/tiny_autoencoder.py:1]: maybe name the file "autoencoder_tiny.py" so it will be with its autoencoders family

File: src/diffusers/models/tiny_autoencoder.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    latents: torch.Tensor
+
+
+class Clamp(nn.Module):
+    def forward(self, x):
+        return torch.tanh(x / 3) * 3
+
+
+class TinyAutoencoderBlock(nn.Module):
+    def __init__(self, in_channels: int, out_channels: int, act_fn: Callable):
```
Comment (yiyixuxu) [inline src/diffusers/models/tiny_autoencoder.py:48]:
```suggestion
    def __init__(self, in_channels: int, out_channels: int, act_fn: str):
```
I think it is better to pass a string around and only do `act_fn = get_activation(act_fn)` where you use it

File: src/diffusers/models/tiny_autoencoder.py [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+        self, x: torch.FloatTensor, return_dict: bool = True
+    ) -> Union[TinyAutoencoderOutput, Tuple[torch.FloatTensor]]:
+        output = self.encoder(x)
+
+        if not return_dict:
+            return (output,)
+
+        return TinyAutoencoderOutput(latents=output)
+
+    apply_forward_hook
```
Comment (yiyixuxu) [inline src/diffusers/models/tiny_autoencoder.py:254]:
```suggestion
    @apply_forward_hook
```

File: src/diffusers/image_processor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yiyixuxu) [reply src/diffusers/image_processor.py:120]: agree with @madebyollin here - should make sure the tiny decoder output is `[-1,1]`.
really no need to create a flag here

**Review by yiyixuxu — APPROVED** [review]
cool!
nice New york cheese cakes 😍😍😍

File: src/diffusers/image_processor.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/image_processor.py]: Done.

**Comment by sayakpaul** [comment]
Thanks @madebyollin and @yiyixuxu!

I think I have addressed all your comments.

File: src/diffusers/models/autoencoder_tiny.py [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        latents (`torch.Tensor`): Encoded outputs of the `Encoder`.
+
+    """
+
+    latents: torch.Tensor
+
+
+class Clamp(nn.Module):
+    def forward(self, x):
+        return torch.tanh(x / 3) * 3
```
Comment (patrickvonplaten) [inline src/diffusers/models/autoencoder_tiny.py:44]:
```suggestion
```

we don't need a class for this IMO

File: src/diffusers/models/autoencoder_tiny.py [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+
+class AutoencoderTinyBlock(nn.Module):
```
Comment (patrickvonplaten) [inline src/diffusers/models/autoencoder_tiny.py:47]: This should go into https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/unet_2d_blocks.py IMo -> we have all blocks collected there

File: src/diffusers/models/autoencoder_tiny.py [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+            if in_channels != out_channels
+            else nn.Identity()
+        )
+        self.fuse = nn.ReLU()
+
+    def forward(self, x):
+        return self.fuse(self.conv(x) + self.skip(x))
+
+
+class EncoderTiny(nn.Module):
```
Comment (patrickvonplaten) [inline src/diffusers/models/autoencoder_tiny.py:69]: can we move this to https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/vae.py

File: src/diffusers/models/autoencoder_tiny.py [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+            else:
+                x = torch.utils.checkpoint.checkpoint(create_custom_forward(self.layers), x)
+
+        else:
+            x = self.layers(x)
+
+        return x
+
+
+class AutoencoderTiny(ModelMixin, ConfigMixin):
```
Comment (patrickvonplaten) [inline src/diffusers/models/autoencoder_tiny.py:171]: IMO this should be the only class in this file

**Review by patrickvonplaten — COMMENTED** [review]
Looks very nice! Can we add some tests here as well? Also it would be great if we could move most of the code into the blocks and vae.py file. The model files in `src/diffusers/models/...` should only have a single class ideally

File: src/diffusers/models/activations.py [hunk 13]
```diff
@@ -8,5 +8,7 @@ def get_activation(act_fn):
         return nn.Mish()
     elif act_fn == "gelu":
         return nn.GELU()
+    elif act_fn == "relu":
```
Comment (patrickvonplaten) [inline src/diffusers/models/activations.py:11]: cool!

File: src/diffusers/models/autoencoder_tiny.py (hunk 9 above)
Comment (patrickvonplaten) [reply src/diffusers/models/autoencoder_tiny.py:44]: Let's just apply `torch.tanh(x / 3) * 3` directly in the code

**Comment by sayakpaul** [comment]
@stevhliu could you check the doc part of the PR and let me know your thoughts?

File: tests/models/test_models_vae.py [hunk 14]
```diff
@@ -223,6 +223,45 @@ def test_forward_with_norm_groups(self):
...
+            "encoder_block_out_channels": (32, 32),
+            "decoder_block_out_channels": (32, 32),
+            "num_encoder_blocks": (1, 2),
+            "num_decoder_blocks": (2, 1),
+        }
+        inputs_dict = self.dummy_input
+        return init_dict, inputs_dict
+
+    def test_outputs_equivalence(self):
+        pass
```
Comment (patrickvonplaten) [inline tests/models/test_models_vae.py:262]: can we also add one integration test where we test that a latent vector is correctly decoded to an image?

**Review by patrickvonplaten — APPROVED** [review]
Looks good to me! Would be nice to add at least one integration test

File: tests/models/test_models_vae.py (hunk 14 above)
Comment (sayakpaul) [reply tests/models/test_models_vae.py:262]: Added in https://github.com/huggingface/diffusers/pull/4384/commits/d7ab16fa6edc56b31c4c1eb3719b369f1fb133a3.

**Comment by sayakpaul** [comment]
@patrickvonplaten I have added an integration test. Works for you? 

@madebyollin I know you're AFK, but as an FYI, I have opened the following:

* https://huggingface.co/madebyollin/taesdxl/discussions/1
* https://huggingface.co/madebyollin/taesd/discussions/1

Once these two PRs are merged, we will merge this PR.

**Review by patrickvonplaten — APPROVED** [review]
Thanks for adding the tests!

File: docs/source/en/_toctree.yml [hunk 15]
```diff
@@ -162,6 +162,8 @@
       title: AutoencoderKL
     - local: api/models/asymmetricautoencoderkl
       title: AsymmetricAutoencoderKL
+    - local: api/models/autoencoder_tiny
+      title: Tiny Autoencoder
```
Comment (stevhliu) [inline docs/source/en/_toctree.yml:166]: To stay consistent with how its spelled in the original repository.

```suggestion
      title: Tiny AutoEncoder
```

**Review by stevhliu — APPROVED** [review]
Nice and clear docstrings! 👏

A small detail, but I found it a little curious that the model is called Tiny AutoEncoder, but its API is `AutoencoderTiny`. Not a big deal, but I think it makes it easier to refer to when they have the same names (ie you don't have to mentally convert Tiny AutoEncoder to `AutoencoderTiny`).

**Comment by sayakpaul** [comment]
> A small detail, but I found it a little curious that the model is called Tiny AutoEncoder, but its API is AutoencoderTiny.

AutoencoderTiny was done to be in line with `AutoencoderKL`.

File: src/diffusers/models/autoencoder_tiny.py [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+
+        self.latent_magnitude = latent_magnitude
+        self.latent_shift = latent_shift
+        self.scaling_factor = scaling_factor
+
+    def _set_gradient_checkpointing(self, module, value=False):
+        if isinstance(module, (EncoderTiny, DecoderTiny)):
+            module.gradient_checkpointing = value
+
+    def scale_latents(self, x):
```
[the remark this answers is not in the document]
Comment (sayakpaul) [reply src/diffusers/models/autoencoder_tiny.py:143]: Since it's not a highlighted method from the docs, I think it's fine as is for now.

File: src/diffusers/models/tiny_autoencoder.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (bghira) [reply src/diffusers/models/tiny_autoencoder.py]: @sayakpaul after my hellscape i've been through with normalisation i'd love for this to be fixed deeper in
